File: sopack/tests/test_book.py
#!/usr/bin/env python3
"""Unit tests for sopack.book — the fixed-seam Book API.

Run:  /workspaces/sdarm/.venv/bin/python3.11 sopack/tests/test_book.py
"""
from __future__ import annotations

import json
import sys
import tempfile
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))  # qdrant/

from sopack import contract
from sopack.book import Block, Book, BookError, dump, load, to_payload, uid, validate


def _plain_sop_book(**overrides) -> Book:
    meta = {
        "book_code": "ABC",
        "lang": "en",
        "book_pair": "ABC",
        "title": "A Book",
        "author": None,
        "year": None,
        "slug": "a-book",
        "corpus": None,
        "page_kind": None,
    }
    meta.update(overrides)
    blocks = [
        Block(para_key="1.1", page=1, para=1, seq=0, chunks=1, text="Hello world today.", words=3),
        Block(para_key="1.2", page=1, para=2, seq=0, chunks=1, text="A second paragraph here.", words=4),
    ]
    return Book(
        schema=contract.SCHEMA_BOOK,
        profile="sop",
        source={"file": "x.json", "sha256": "abc", "kind": "sop_json",
                "acquired_from": None, "rights": None},
        book=meta,
        id_rule="sop/plain",
        alignment=None,
        stats={"blocks_in": 2, "blocks_out": 2, "dropped": 0, "damage": 0.0,
               "words": 7, "dropped_detail": []},
        blocks=blocks,
    )


class DumpLoadRoundTrip(unittest.TestCase):
    def test_round_trip_preserves_content(self):
        book = _plain_sop_book()
        with tempfile.TemporaryDirectory() as td:
            path = Path(td) / "book.json"
            dump(book, path)
            self.assertTrue(path.exists())
            back = load(path)
        self.assertEqual(back.schema, book.schema)
        self.assertEqual(back.profile, book.profile)
        self.assertEqual(back.book, book.book)
        self.assertEqual(back.id_rule, book.id_rule)
        self.assertEqual(back.alignment, book.alignment)
        self.assertEqual(back.stats, book.stats)
        self.assertEqual(back.blocks, book.blocks)

    def test_dump_has_stable_key_order_and_indent(self):
        book = _plain_sop_book()
        with tempfile.TemporaryDirectory() as td:
            path = Path(td) / "book.json"
            dump(book, path)
            raw = path.read_text(encoding="utf-8")
        doc = json.loads(raw)
        self.assertEqual(list(doc.keys()),
                          ["schema", "profile", "source", "book", "id_rule",
                           "alignment", "stats", "blocks"])
        self.assertIn("\n  ", raw)  # indent=2
        block_keys = list(doc["blocks"][0].keys())
        self.assertEqual(block_keys,
                          ["para_key", "page", "para", "seq", "chunks", "text", "words"])

    def test_load_rejects_bad_json(self):
        with tempfile.TemporaryDirectory() as td:
            path = Path(td) / "book.json"
            path.write_text("{not json", encoding="utf-8")
            with self.assertRaises(BookError):
                load(path)

    def test_load_rejects_missing_top_level_key(self):
        with tempfile.TemporaryDirectory() as td:
            path = Path(td) / "book.json"
            doc = {"schema": contract.SCHEMA_BOOK, "profile": "sop"}
            path.write_text(json.dumps(doc), encoding="utf-8")
            with self.assertRaises(BookError):
                load(path)

    def test_load_rejects_unknown_profile(self):
        with tempfile.TemporaryDirectory() as td:
            path = Path(td) / "book.json"
            book = _plain_sop_book(book_code=None)
            dump(book, path)
            doc = json.loads(path.read_text(encoding="utf-8"))
            doc["profile"] = "nonsense"
            path.write_text(json.dumps(doc), encoding="utf-8")
            with self.assertRaises(BookError):
                load(path)

    def test_load_rejects_malformed_block(self):
        with tempfile.TemporaryDirectory() as td:
            path = Path(td) / "book.json"
            book = _plain_sop_book()
            dump(book, path)
            doc = json.loads(path.read_text(encoding="utf-8"))
            del doc["blocks"][0]["words"]
            path.write_text(json.dumps(doc), encoding="utf-8")
            with self.assertRaises(BookError):
                load(path)

    def test_load_accepts_null_metadata(self):
        """extract never invents metadata (rule #9) — load() must not choke
        on a book.json whose book_code/author/etc. are still null; that is
        validate()'s job to report, not load()'s job to reject."""
        book = _plain_sop_book(book_code=None, title=None)
        with tempfile.TemporaryDirectory() as td:
            path = Path(td) / "book.json"
            dump(book, path)
            back = load(path)  # must not raise
        self.assertIsNone(back.book["book_code"])


class Validate(unittest.TestCase):
    def test_valid_book_has_no_errors(self):
        self.assertEqual(validate(_plain_sop_book()), [])

    def test_missing_required_metadata(self):
        book = _plain_sop_book(book_code=None, lang=None, title=None)
        errors = validate(book)
        self.assertTrue(any("book_code" in e for e in errors))
        self.assertTrue(any("lang" in e for e in errors))
        self.assertTrue(any("title" in e for e in errors))

    def test_egw_exempt_from_author_year(self):
        # corpus is None -> treated as EGW -> author/year not required
        book = _plain_sop_book(corpus=None, author=None, year=None)
        errors = validate(book)
        self.assertFalse(any("author" in e for e in errors))
        self.assertFalse(any("year" in e for e in errors))

    def test_non_egw_requires_author_year(self):
        book = _plain_sop_book(corpus="pioneers", author=None, year=None)
        errors = validate(book)
        self.assertTrue(any("author" in e for e in errors))
        self.assertTrue(any("year" in e for e in errors))

    def test_non_egw_with_author_year_is_clean(self):
        book = _plain_sop_book(corpus="pioneers", author="J. N. Andrews", year=1873)
        self.assertEqual(validate(book), [])

    def test_rejects_bad_id_rule_for_profile(self):
        book = _plain_sop_book()
        book.id_rule = "bible/v1"
        errors = validate(book)
        self.assertTrue(any("id_rule" in e for e in errors))

    def test_rejects_duplicate_para_key_seq(self):
        book = _plain_sop_book()
        book.blocks.append(Block(para_key="1.1", page=1, para=1, seq=0, chunks=1,
                                  text="duplicate", words=1))
        errors = validate(book)
        self.assertTrue(any("duplicate" in e for e in errors))

    def test_rejects_empty_text(self):
        book = _plain_sop_book()
        book.blocks[0].text = "   "
        errors = validate(book)
        self.assertTrue(any("empty text" in e for e in errors))

    def test_rejects_chunks_seq_inconsistency_gap(self):
        book = _plain_sop_book()
        # A split block that skips seq 1 (claims chunks=3 but only 0 and 2 exist)
        book.blocks = [
            Block(para_key="9.1", page=9, para=1, seq=0, chunks=3, text="part one here", words=3),
            Block(para_key="9.1", page=9, para=1, seq=2, chunks=3, text="part three here", words=3),
        ]
        errors = validate(book)
        self.assertTrue(any("9.1" in e and "seq" in e for e in errors))

    def test_rejects_inconsistent_chunks_value(self):
        book = _plain_sop_book()
        book.blocks = [
            Block(para_key="9.1", page=9, para=1, seq=0, chunks=2, text="part one here", words=3),
            Block(para_key="9.1", page=9, para=1, seq=1, chunks=3, text="part two here", words=3),
        ]
        errors = validate(book)
        self.assertTrue(any("inconsistent" in e for e in errors))

    def test_valid_chunked_block_is_clean(self):
        book = _plain_sop_book()
        book.blocks = [
            Block(para_key="9.1", page=9, para=1, seq=0, chunks=2, text="part one here", words=3),
            Block(para_key="9.1", page=9, para=1, seq=1, chunks=2, text="part two here", words=3),
        ]
        self.assertEqual(validate(book), [])


class Payload(unittest.TestCase):
    def test_sop_payload_matches_contract(self):
        book = _plain_sop_book()
        profile = contract.get_profile("sop")
        for block in book.blocks:
            payload = to_payload(book, block)
            self.assertEqual(contract.validate_payload(profile, payload), [])

    def test_sop_payload_carries_required_keys_even_when_none(self):
        book = _plain_sop_book(book_pair=None)
        payload = to_payload(book, book.blocks[0])
        self.assertIn("book_pair", payload)
        self.assertIsNone(payload["book_pair"])
        self.assertIn("aligned", payload)
        self.assertIsNone(payload["aligned"])

    def test_sop_payload_omits_absent_optional_keys(self):
        book = _plain_sop_book(author=None, title="Some Title", year=None, slug=None,
                                page_kind=None, corpus=None)
        payload = to_payload(book, book.blocks[0])
        self.assertNotIn("author", payload)
        self.assertNotIn("year", payload)
        self.assertNotIn("slug", payload)
        self.assertNotIn("page_kind", payload)
        self.assertNotIn("corpus", payload)
        self.assertEqual(payload["title"], "Some Title")

    def test_sop_payload_chunk_fields_only_when_split(self):
        book = _plain_sop_book()
        unsplit = Block(para_key="2.1", page=2, para=1, seq=0, chunks=1, text="one piece", words=2)
        split = Block(para_key="2.2", page=2, para=2, seq=1, chunks=2, text="second piece", words=2)
        p1 = to_payload(book, unsplit)
        p2 = to_payload(book, split)
        self.assertNotIn("chunk", p1)
        self.assertNotIn("chunks", p1)
        self.assertEqual(p2["chunk"], 1)
        self.assertEqual(p2["chunks"], 2)

    def test_aligned_pulled_from_alignment_en_reverse(self):
        # en_reverse is keyed by EN para_key -> [own-language para_keys] (the
        # sop_json shape); a DE book inverts it to go from its own para_key
        # back to the aligned EN para_key(s).
        book = _plain_sop_book(lang="de", book_code="BH")
        book.alignment = {"en_code": "SL", "en_reverse": {"7.1": ["1.1"]}}
        payload = to_payload(book, book.blocks[0])  # para_key "1.1"
        self.assertEqual(payload["aligned"], ["7.1"])
        payload2 = to_payload(book, book.blocks[1])  # "1.2" has no alignment entry
        self.assertIsNone(payload2["aligned"])

    def test_aligned_direct_lookup_for_english_book(self):
        book = _plain_sop_book(lang="en", book_code="SL")
        book.alignment = {"en_code": "SL", "en_reverse": {"1.1": ["5.1"]}}
        payload = to_payload(book, book.blocks[0])  # para_key "1.1"
        self.assertEqual(payload["aligned"], ["5.1"])

    def test_bible_payload_matches_contract(self):
        book = _plain_sop_book(book_code="KJV")
        book.profile = "bible"
        profile = contract.get_profile("bible")
        payload = to_payload(book, book.blocks[0])
        self.assertEqual(contract.validate_payload(profile, payload), [])


class Uid(unittest.TestCase):
    def test_uid_agrees_with_contract_for_sop_plain(self):
        book = _plain_sop_book(lang="en", book_code="ABC")
        block = book.blocks[0]
        got = uid(book, block)
        want = contract.uid_for("sop/plain", {"lang": "en", "book_code": "ABC", "para_key": "1.1"})
        self.assertEqual(got, want)

    def test_uid_agrees_with_contract_for_sop_seq(self):
        book = _plain_sop_book(lang="en", book_code="ABC")
        book.id_rule = "sop/seq"
        block = Block(para_key="3.4", page=3, para=4, seq=2, chunks=3, text="x y z", words=3)
        got = uid(book, block)
        want = contract.uid_for("sop/seq",
                                 {"lang": "en", "book_code": "ABC", "para_key": "3.4", "seq": 2})
        self.assertEqual(got, want)

    def test_uid_agrees_with_contract_for_bible(self):
        book = _plain_sop_book(book_code="KJV")
        book.profile = "bible"
        book.id_rule = "bible/v1"
        block = book.blocks[0]
        got = uid(book, block)
        want = contract.uid_for("bible/v1", {"bible": "KJV", "osis": block.para_key})
        self.assertEqual(got, want)

    def test_uid_feeds_a_stable_point_id(self):
        book = _plain_sop_book(lang="en", book_code="ABC")
        pid1 = contract.point_id(book.id_rule, {"lang": "en", "book_code": "ABC", "para_key": "1.1"})
        pid2 = contract.point_id(book.id_rule, {"lang": "en", "book_code": "ABC", "para_key": "1.1"})
        self.assertEqual(pid1, pid2)


if __name__ == "__main__":
    unittest.main()

File: sopack-rs/crates/sopack-book/tests/test_book.rs
//! Port of `qdrant/sopack/tests/test_book.py` — same test names (minus the
//! `test_` prefix's Python-class grouping, folded into Rust module names),
//! same behaviour asserted.

use std::collections::HashMap;

use serde_json::{json, Map, Value};
use sopack_book::{
    dump, get_profile, load, point_id, to_payload, uid, uid_for, validate, validate_payload, Block,
    Book, SCHEMA_BOOK,
};

fn obj(v: Value) -> Map<String, Value> {
    match v {
        Value::Object(m) => m,
        _ => panic!("expected object"),
    }
}

/// Port of `_plain_sop_book()`. `overrides` lets a caller replace individual
/// `book` metadata keys, mirroring the Python `**overrides` kwarg.
fn plain_sop_book(overrides: &[(&str, Value)]) -> Book {
    let mut meta = obj(json!({
        "book_code": "ABC",
        "lang": "en",
        "book_pair": "ABC",
        "title": "A Book",
        "author": null,
        "year": null,
        "slug": "a-book",
        "corpus": null,
        "page_kind": null,
    }));
    for (k, v) in overrides {
        meta.insert((*k).to_string(), v.clone());
    }
    let blocks = vec![
        Block::new("1.1", 1, 1, 0, 1, "Hello world today.", 3, None),
        Block::new("1.2", 1, 2, 0, 1, "A second paragraph here.", 4, None),
    ];
    Book {
        schema: SCHEMA_BOOK.to_string(),
        profile: "sop".to_string(),
        source: obj(json!({
            "file": "x.json", "sha256": "abc", "kind": "sop_json",
            "acquired_from": null, "rights": null,
        })),
        book: meta,
        id_rule: "sop/plain".to_string(),
        alignment: None,
        stats: obj(json!({
            "blocks_in": 2, "blocks_out": 2, "dropped": 0, "damage": 0.0,
            "words": 7, "dropped_detail": [],
        })),
        blocks,
    }
}

fn tmp_path(name: &str) -> std::path::PathBuf {
    let dir = tempfile::tempdir().expect("tempdir");
    let path = dir.path().join(name);
    // Leak the tempdir so the file survives for the caller; tests are
    // short-lived processes so this is fine.
    std::mem::forget(dir);
    path
}

// ── DumpLoadRoundTrip ────────────────────────────────────────────────────

#[test]
fn test_round_trip_preserves_content() {
    let book = plain_sop_book(&[]);
    let path = tmp_path("book.json");
    dump(&book, &path).unwrap();
    assert!(path.exists());
    let back = load(&path).unwrap();
    assert_eq!(back.schema, book.schema);
    assert_eq!(back.profile, book.profile);
    assert_eq!(back.book, book.book);
    assert_eq!(back.id_rule, book.id_rule);
    assert_eq!(back.alignment, book.alignment);
    assert_eq!(back.stats, book.stats);
    assert_eq!(back.blocks, book.blocks);
}

#[test]
fn test_dump_has_stable_key_order_and_indent() {
    let book = plain_sop_book(&[]);
    let path = tmp_path("book.json");
    dump(&book, &path).unwrap();
    let raw = std::fs::read_to_string(&path).unwrap();
    let doc: Value = serde_json::from_str(&raw).unwrap();
    let keys: Vec<&String> = doc.as_object().unwrap().keys().collect();
    assert_eq!(
        keys,
        vec![
            "schema",
            "profile",
            "source",
            "book",
            "id_rule",
            "alignment",
            "stats",
            "blocks"
        ]
    );
    assert!(raw.contains("\n  "));
    let block_keys: Vec<&String> = doc["blocks"][0].as_object().unwrap().keys().collect();
    assert_eq!(
        block_keys,
        vec!["para_key", "page", "para", "seq", "chunk", "chunks", "text", "words"]
    );
}

#[test]
fn test_load_defaults_chunk_to_seq_for_pre_0_1_4_files() {
    let book = plain_sop_book(&[]);
    let path = tmp_path("book.json");
    dump(&book, &path).unwrap();
    let raw = std::fs::read_to_string(&path).unwrap();
    let mut doc: Value = serde_json::from_str(&raw).unwrap();
    for b in doc["blocks"].as_array_mut().unwrap() {
        b.as_object_mut().unwrap().remove("chunk");
    }
    doc["blocks"][1]["seq"] = json!(1);
    doc["blocks"][1]["para_key"] = json!("1.1");
    std::fs::write(&path, serde_json::to_string(&doc).unwrap()).unwrap();
    let loaded = load(&path).unwrap();
    let chunks: Vec<i64> = loaded.blocks.iter().map(|b| b.chunk).collect();
    assert_eq!(chunks, vec![0, 1]);
}

#[test]
fn test_dump_load_round_trips_chunk() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![
        Block::new("3.1", 3, 1, 0, 1, "a heading", 2, Some(0)),
        Block::new("3.1", 3, 1, 1, 2, "piece one", 2, Some(0)),
        Block::new("3.1", 3, 1, 2, 2, "piece two", 2, Some(1)),
    ];
    let path = tmp_path("book.json");
    dump(&book, &path).unwrap();
    let loaded = load(&path).unwrap();
    let got: Vec<(i64, i64)> = loaded.blocks.iter().map(|b| (b.seq, b.chunk)).collect();
    assert_eq!(got, vec![(0, 0), (1, 0), (2, 1)]);
    assert_eq!(validate(&loaded), Vec::<String>::new());
}

#[test]
fn test_load_rejects_bad_json() {
    let path = tmp_path("book.json");
    std::fs::write(&path, "{not json").unwrap();
    assert!(load(&path).is_err());
}

#[test]
fn test_load_rejects_missing_top_level_key() {
    let path = tmp_path("book.json");
    let doc = json!({"schema": SCHEMA_BOOK, "profile": "sop"});
    std::fs::write(&path, serde_json::to_string(&doc).unwrap()).unwrap();
    assert!(load(&path).is_err());
}

#[test]
fn test_load_rejects_unknown_profile() {
    let book = plain_sop_book(&[("book_code", Value::Null)]);
    let path = tmp_path("book.json");
    dump(&book, &path).unwrap();
    let raw = std::fs::read_to_string(&path).unwrap();
    let mut doc: Value = serde_json::from_str(&raw).unwrap();
    doc["profile"] = json!("nonsense");
    std::fs::write(&path, serde_json::to_string(&doc).unwrap()).unwrap();
    assert!(load(&path).is_err());
}

#[test]
fn test_load_rejects_malformed_block() {
    let book = plain_sop_book(&[]);
    let path = tmp_path("book.json");
    dump(&book, &path).unwrap();
    let raw = std::fs::read_to_string(&path).unwrap();
    let mut doc: Value = serde_json::from_str(&raw).unwrap();
    doc["blocks"][0].as_object_mut().unwrap().remove("words");
    std::fs::write(&path, serde_json::to_string(&doc).unwrap()).unwrap();
    assert!(load(&path).is_err());
}

#[test]
fn test_load_accepts_null_metadata() {
    let book = plain_sop_book(&[("book_code", Value::Null), ("title", Value::Null)]);
    let path = tmp_path("book.json");
    dump(&book, &path).unwrap();
    let back = load(&path).unwrap(); // must not raise
    assert!(back.book.get("book_code").unwrap().is_null());
}

// ── Validate ─────────────────────────────────────────────────────────────

#[test]
fn test_valid_book_has_no_errors() {
    assert_eq!(validate(&plain_sop_book(&[])), Vec::<String>::new());
}

#[test]
fn test_missing_required_metadata() {
    let book = plain_sop_book(&[
        ("book_code", Value::Null),
        ("lang", Value::Null),
        ("title", Value::Null),
    ]);
    let errors = validate(&book);
    assert!(errors.iter().any(|e| e.contains("book_code")));
    assert!(errors.iter().any(|e| e.contains("lang")));
    assert!(errors.iter().any(|e| e.contains("title")));
}

#[test]
fn test_egw_exempt_from_author_year() {
    let book = plain_sop_book(&[
        ("corpus", Value::Null),
        ("author", Value::Null),
        ("year", Value::Null),
    ]);
    let errors = validate(&book);
    assert!(!errors.iter().any(|e| e.contains("author")));
    assert!(!errors.iter().any(|e| e.contains("year")));
}

#[test]
fn test_non_egw_requires_author_year() {
    let book = plain_sop_book(&[
        ("corpus", json!("pioneers")),
        ("author", Value::Null),
        ("year", Value::Null),
    ]);
    let errors = validate(&book);
    assert!(errors.iter().any(|e| e.contains("author")));
    assert!(errors.iter().any(|e| e.contains("year")));
}

#[test]
fn test_non_egw_with_author_year_is_clean() {
    let book = plain_sop_book(&[
        ("corpus", json!("pioneers")),
        ("author", json!("J. N. Andrews")),
        ("year", json!(1873)),
    ]);
    assert_eq!(validate(&book), Vec::<String>::new());
}

#[test]
fn test_rejects_bad_id_rule_for_profile() {
    let mut book = plain_sop_book(&[]);
    book.id_rule = "bible/v1".to_string();
    let errors = validate(&book);
    assert!(errors.iter().any(|e| e.contains("id_rule")));
}

#[test]
fn test_rejects_duplicate_para_key_seq() {
    let mut book = plain_sop_book(&[]);
    book.blocks
        .push(Block::new("1.1", 1, 1, 0, 1, "duplicate", 1, None));
    let errors = validate(&book);
    assert!(errors.iter().any(|e| e.contains("duplicate")));
}

#[test]
fn test_rejects_empty_text() {
    let mut book = plain_sop_book(&[]);
    book.blocks[0].text = "   ".to_string();
    let errors = validate(&book);
    assert!(errors.iter().any(|e| e.contains("empty text")));
}

#[test]
fn test_rejects_chunks_seq_inconsistency_gap() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![
        Block::new("9.1", 9, 1, 0, 3, "part one here", 3, None),
        Block::new("9.1", 9, 1, 2, 3, "part three here", 3, None),
    ];
    let errors = validate(&book);
    assert!(errors
        .iter()
        .any(|e| e.contains("9.1") && e.contains("seq")));
}

#[test]
fn test_rejects_inconsistent_chunks_value() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![
        Block::new("9.1", 9, 1, 0, 2, "part one here", 3, None),
        Block::new("9.1", 9, 1, 1, 3, "part two here", 3, None),
    ];
    let errors = validate(&book);
    assert!(errors.iter().any(|e| e.contains("inconsistent")));
}

#[test]
fn test_accepts_two_paragraphs_sharing_one_para_key() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![
        Block::new("3.1", 3, 1, 0, 1, "Revelation 18 verse one", 4, Some(0)),
        Block::new(
            "3.1",
            3,
            1,
            1,
            1,
            "II. WHAT ARE WE TO UNDERSTAND",
            6,
            Some(0),
        ),
    ];
    assert_eq!(validate(&book), Vec::<String>::new());
}

#[test]
fn test_accepts_split_paragraph_after_a_colliding_one() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![
        Block::new("3.1", 3, 1, 0, 1, "the heading here", 3, Some(0)),
        Block::new("3.1", 3, 1, 1, 2, "a long paragraph", 3, Some(0)),
        Block::new("3.1", 3, 1, 2, 2, "split in two", 3, Some(1)),
    ];
    assert_eq!(validate(&book), Vec::<String>::new());
}

#[test]
fn test_rejects_seq_gap_across_colliding_paragraphs() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![
        Block::new("3.1", 3, 1, 0, 1, "first here", 2, Some(0)),
        Block::new("3.1", 3, 1, 2, 1, "third here", 2, Some(0)),
    ];
    let errors = validate(&book);
    assert!(errors
        .iter()
        .any(|e| e.contains("3.1") && e.contains("seq")));
}

#[test]
fn test_rejects_truncated_split_paragraph() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![Block::new("3.1", 3, 1, 0, 2, "only piece", 2, Some(0))];
    let errors = validate(&book);
    assert!(errors.iter().any(|e| e.contains("chunks=2")));
}

#[test]
fn test_valid_chunked_block_is_clean() {
    let mut book = plain_sop_book(&[]);
    book.blocks = vec![
        Block::new("9.1", 9, 1, 0, 2, "part one here", 3, None),
        Block::new("9.1", 9, 1, 1, 2, "part two here", 3, None),
    ];
    assert_eq!(validate(&book), Vec::<String>::new());
}

// ── Payload ──────────────────────────────────────────────────────────────

#[test]
fn test_sop_payload_matches_contract() {
    let book = plain_sop_book(&[]);
    let profile = get_profile("sop").unwrap();
    for block in &book.blocks {
        let payload = to_payload(&book, block).unwrap();
        assert_eq!(validate_payload(profile, &payload), Vec::<String>::new());
    }
}

#[test]
fn test_sop_payload_carries_required_keys_even_when_none() {
    let book = plain_sop_book(&[("book_pair", Value::Null)]);
    let payload = to_payload(&book, &book.blocks[0]).unwrap();
    assert!(payload.contains_key("book_pair"));
    assert!(payload["book_pair"].is_null());
    assert!(payload.contains_key("aligned"));
    assert!(payload["aligned"].is_null());
}

#[test]
fn test_sop_payload_omits_absent_optional_keys() {
    let book = plain_sop_book(&[
        ("author", Value::Null),
        ("title", json!("Some Title")),
        ("year", Value::Null),
        ("slug", Value::Null),
        ("page_kind", Value::Null),
        ("corpus", Value::Null),
    ]);
    let payload = to_payload(&book, &book.blocks[0]).unwrap();
    assert!(!payload.contains_key("author"));
    assert!(!payload.contains_key("year"));
    assert!(!payload.contains_key("slug"));
    assert!(!payload.contains_key("page_kind"));
    assert!(!payload.contains_key("corpus"));
    assert_eq!(payload["title"], json!("Some Title"));
}

#[test]
fn test_sop_payload_chunk_fields_only_when_split() {
    let book = plain_sop_book(&[]);
    let unsplit = Block::new("2.1", 2, 1, 0, 1, "one piece", 2, None);
    let split = Block::new("2.2", 2, 2, 1, 2, "second piece", 2, None);
    let p1 = to_payload(&book, &unsplit).unwrap();
    let p2 = to_payload(&book, &split).unwrap();
    assert!(!p1.contains_key("chunk"));
    assert!(!p1.contains_key("chunks"));
    assert_eq!(p2["chunk"], json!(1));
    assert_eq!(p2["chunks"], json!(2));
}

#[test]
fn test_sop_payload_chunk_is_paragraph_local_not_seq() {
    let book = plain_sop_book(&[]);
    let block = Block::new("3.1", 3, 1, 1, 2, "first piece", 2, Some(0));
    let payload = to_payload(&book, &block).unwrap();
    assert_eq!(payload["chunk"], json!(0));
    assert_eq!(payload["chunks"], json!(2));
}

#[test]
fn test_aligned_pulled_from_alignment_en_reverse() {
    let mut book = plain_sop_book(&[("lang", json!("de")), ("book_code", json!("BH"))]);
    book.alignment = Some(obj(json!({
        "en_code": "SL",
        "en_reverse": {"7.1": ["1.1"]},
    })));
    let payload = to_payload(&book, &book.blocks[0]).unwrap(); // para_key "1.1"
    assert_eq!(payload["aligned"], json!(["7.1"]));
    let payload2 = to_payload(&book, &book.blocks[1]).unwrap(); // "1.2" has no entry
    assert!(payload2["aligned"].is_null());
}

#[test]
fn test_aligned_direct_lookup_for_english_book() {
    let mut book = plain_sop_book(&[("lang", json!("en")), ("book_code", json!("SL"))]);
    book.alignment = Some(obj(json!({
        "en_code": "SL",
        "en_reverse": {"1.1": ["5.1"]},
    })));
    let payload = to_payload(&book, &book.blocks[0]).unwrap(); // para_key "1.1"
    assert_eq!(payload["aligned"], json!(["5.1"]));
}

#[test]
fn test_bible_payload_matches_contract() {
    let mut book = plain_sop_book(&[("book_code", json!("KJV"))]);
    book.profile = "bible".to_string();
    let profile = get_profile("bible").unwrap();
    let payload = to_payload(&book, &book.blocks[0]).unwrap();
    assert_eq!(validate_payload(profile, &payload), Vec::<String>::new());
}

// ── Uid ──────────────────────────────────────────────────────────────────

fn fields(pairs: &[(&'static str, Option<&str>)]) -> HashMap<&'static str, Option<String>> {
    let mut m = HashMap::new();
    for (k, v) in pairs {
        m.insert(*k, v.map(str::to_string));
    }
    m
}

#[test]
fn test_uid_agrees_with_contract_for_sop_plain() {
    let book = plain_sop_book(&[("lang", json!("en")), ("book_code", json!("ABC"))]);
    let block = &book.blocks[0];
    let got = uid(&book, block).unwrap();
    let want = uid_for(
        "sop/plain",
        &fields(&[
            ("lang", Some("en")),
            ("book_code", Some("ABC")),
            ("para_key", Some("1.1")),
        ]),
    )
    .unwrap();
    assert_eq!(got, want);
}

#[test]
fn test_uid_agrees_with_contract_for_sop_seq() {
    let mut book = plain_sop_book(&[("lang", json!("en")), ("book_code", json!("ABC"))]);
    book.id_rule = "sop/seq".to_string();
    let block = Block::new("3.4", 3, 4, 2, 3, "x y z", 3, None);
    let got = uid(&book, &block).unwrap();
    let want = uid_for(
        "sop/seq",
        &fields(&[
            ("lang", Some("en")),
            ("book_code", Some("ABC")),
            ("para_key", Some("3.4")),
            ("seq", Some("2")),
        ]),
    )
    .unwrap();
    assert_eq!(got, want);
}

#[test]
fn test_uid_agrees_with_contract_for_bible() {
    let mut book = plain_sop_book(&[("book_code", json!("KJV"))]);
    book.profile = "bible".to_string();
    book.id_rule = "bible/v1".to_string();
    let block = book.blocks[0].clone();
    let got = uid(&book, &block).unwrap();
    let want = uid_for(
        "bible/v1",
        &fields(&[
            ("bible", Some("KJV")),
            ("osis", Some(block.para_key.as_str())),
        ]),
    )
    .unwrap();
    assert_eq!(got, want);
}

#[test]
fn test_uid_feeds_a_stable_point_id() {
    let book = plain_sop_book(&[("lang", json!("en")), ("book_code", json!("ABC"))]);
    let f = fields(&[
        ("lang", Some("en")),
        ("book_code", Some("ABC")),
        ("para_key", Some("1.1")),
    ]);
    let pid1 = point_id(&book.id_rule, &f).unwrap();
    let pid2 = point_id(&book.id_rule, &f).unwrap();
    assert_eq!(pid1, pid2);
}
